web/src/test/java/org/springframework/security/web/csrf/XorCsrfTokenRepositoryTests.java

```diff
@@ -0,0 +1,175 @@
+		given(this.delegate.generateToken(this.request)).willReturn(this.csrfToken);
+
+		CsrfToken csrfToken = this.repository.generateToken(this.request);
+		assertThat(csrfToken).isNotInstanceOf(DefaultCsrfToken.class);
+		assertThat(csrfToken.getHeaderName()).isEqualTo(this.csrfToken.getHeaderName());
+		assertThat(csrfToken.getParameterName()).isEqualTo(this.csrfToken.getParameterName());
+
+		String token1 = csrfToken.getToken();
+		String token2 = csrfToken.getToken();
+		assertThat(token1).isNotEqualTo(token2);
+		assertThat(csrfToken.matches(token1)).isTrue();
+		assertThat(csrfToken.matches(token2)).isTrue();
+
+		verify(this.delegate).generateToken(this.request);
+	}
+
+	@Test
+	public void saveTokenWhenNullThenCallsDelegateWithNull() {
+		this.repository.saveToken(null, this.request, this.response);
+		verify(this.delegate).saveToken(null, this.request, this.response);
+	}
+
+	@Test
+	public void saveTokenWhenDefaultCsrfTokenThenCallsDelegateWithDefaultCsrfToken() {
+		this.repository.saveToken(this.csrfToken, this.request, this.response);
+		verify(this.delegate).saveToken(this.csrfToken, this.request, this.response);
+	}
+
+	@Test
+	public void saveTokenWhenWrappedCsrfTokenThenCallsDelegateWithDefaultCsrfToken() {
+		given(this.delegate.generateToken(this.request)).willReturn(this.csrfToken);
+
+		CsrfToken csrfToken = this.repository.generateToken(this.request);
+		assertThat(csrfToken).isNotInstanceOf(DefaultCsrfToken.class);
+		this.repository.saveToken(csrfToken, this.request, this.response);
+
+		verify(this.delegate).generateToken(this.request);
+		verify(this.delegate).saveToken(this.csrfToken, this.request, this.response);
+	}
+
+	@Test
+	public void loadTokenWhenCalledThenReturnsWrappedCsrfToken() {
+		given(this.delegate.loadToken(this.request)).willReturn(this.csrfToken);
+
+		CsrfToken csrfToken = this.repository.loadToken(this.request);
+		assertThat(csrfToken).isNotInstanceOf(DefaultCsrfToken.class);
+
+		String token1 = csrfToken.getToken();
+		String token2 = csrfToken.getToken();
+		assertThat(token1).isNotEqualTo(token2);
+
+		verify(this.delegate).loadToken(this.request);
+	}
+
+	@Test
+	public void matchesWhenTokensEqualThenTrue() {
+		given(this.delegate.loadToken(this.request)).willReturn(this.csrfToken);
+
+		CsrfToken csrfToken1 = this.repository.loadToken(this.request);
+		CsrfToken csrfToken2 = this.repository.loadToken(this.request);
+		String token1 = csrfToken1.getToken();
+		String token2 = csrfToken2.getToken();
+		assertThat(token1).isNotEqualTo(token2);
+		assertThat(csrfToken1.matches(token1)).isTrue();
+		assertThat(csrfToken1.matches(token2)).isTrue();
+		assertThat(csrfToken2.matches(token1)).isTrue();
+		assertThat(csrfToken2.matches(token2)).isTrue();
+	}
+
+	@Test
+	public void matchesWhenTokensNotEqualThenFalse() {
+		given(this.delegate.generateToken(this.request)).willAnswer(
+				(invocation) -> new DefaultCsrfToken("headerName", "parameterName", UUID.randomUUID().toString()));
```

Comment:
Avoid use of random as much as possible because it slows the tests down and we want tests to be consistent. If you want a random value, generate one and copy the value into the test.